From b8e47f68a10c261d5b24c3076f90187ba0f538a3 Mon Sep 17 00:00:00 2001
From: "David C. Manuelda" <StormByte@gmail.com>
Date: Mon, 12 Oct 2026 18:57:18 +0200
Subject: [PATCH] meson: gate NVIDIA codecs by platform support

Disable NVIDIA codec auto-detection on unsupported targets and skip the related pkg-config, type, and compiler probes. Report a targeted error when ffnvcodec is explicitly requested on an unsupported platform.

Select NVCC automatically for the CUDA SDK only when CUDA LLVM was not explicitly enabled, while respecting an explicitly disabled NVCC option.
---
 meson.build | 69 +++++++++++++++++++++++++++++++++++++++++++++++++----
 1 file changed, 64 insertions(+), 5 deletions(-)

diff --git a/meson.build b/meson.build
index b2898afa9c..cb49fdfe8d 100644
--- a/meson.build
+++ b/meson.build
@@ -196,6 +196,23 @@ common_deps = [thread_dep]
 cpu = host_machine.cpu_family()
 host_system = host_machine.system()
 
+# Match configure's supported architectures and operating systems for the
+# dynamically loaded NVIDIA codec APIs.
+nvidia_codec_supported = false
+if ['x86', 'x86_64'].contains(cpu)
+  nvidia_codec_supported = ['linux', 'freebsd', 'windows', 'cygwin'].contains(host_system)
+elif ['aarch64', 'ppc64'].contains(cpu) and host_machine.endian() == 'little'
+  nvidia_codec_supported = ['linux', 'windows'].contains(host_system)
+endif
+
+ffnvcodec_opt = get_option('ffnvcodec').disable_auto_if(not nvidia_codec_supported).disable_if(
+  not nvidia_codec_supported,
+  error_message: 'ffnvcodec is not supported on @0@/@1@'.format(host_system, cpu),
+)
+nvenc_opt = get_option('nvenc').disable_auto_if(not nvidia_codec_supported)
+nvdec_opt = get_option('nvdec').disable_auto_if(not nvidia_codec_supported)
+cuvid_opt = get_option('cuvid').disable_auto_if(not nvidia_codec_supported)
+
 if host_system == 'windows'
   common_deps += [cc.find_library('ws2_32')]
 endif
@@ -203,6 +220,13 @@ endif
 conf = configuration_data()
 conf_requested = configuration_data()
 
+if not nvidia_codec_supported
+  conf.set10('ffnvcodec', false)
+  conf.set10('nvenc', false)
+  conf.set10('nvdec', false)
+  conf.set10('cuvid', false)
+endif
+
 compat_objs = []
 
 document_list = [
@@ -980,8 +1004,6 @@ check_pkg_config = [
   ['alsa', ['alsa/asoundlib.h'], ['snd_pcm_htimestamp'], {}],
   ['cairo', ['cairo.h'], ['cairo_create'], {'pkg': 'cairo'}],
   ['chromaprint', ['chromaprint.h'], ['chromaprint_get_version'], {'pkg': 'chromaprint'}],
-  ['ffnvcodec', ['ffnvcodec/nvEncodeAPI.h', 'ffnvcodec/dynlink_cuda.h', 'ffnvcodec/dynlink_cuviddec.h', 'ffnvcodec/dynlink_nvcuvid.h'],
-    [], {'versions': ['>= 12.1.14.0', ['>= 12.0.16.1', '< 12.1'], ['>= 11.1.5.3', '< 12.0']]}],
   ['libgme', ['gme/gme.h'], ['gme_new_emu'], {}],
   ['gnutls', ['gnutls/gnutls.h'], ['gnutls_global_init'], {}],
   ['lcms2', ['lcms2.h'], ['cmsCreateContext'], {'pkg': 'lcms2', 'version': '>= 2.13'}],
@@ -1087,6 +1109,13 @@ check_pkg_config = [
   ['bzlib', ['bzlib.h'], ['BZ2_bzlibVersion'], {'pkg': 'bzip2'}],
 ]
 
+if ffnvcodec_opt.allowed()
+  check_pkg_config += [
+    ['ffnvcodec', ['ffnvcodec/nvEncodeAPI.h', 'ffnvcodec/dynlink_cuda.h', 'ffnvcodec/dynlink_cuviddec.h', 'ffnvcodec/dynlink_nvcuvid.h'],
+      [], {'versions': ['>= 12.1.14.0', ['>= 12.0.16.1', '< 12.1'], ['>= 11.1.5.3', '< 12.0']]}],
+  ]
+endif
+
 if get_option('xlib').allowed()
   check_pkg_config += [
     ['xlib_x11', ['X11/Xlib.h'], ['XPending'], {'pkg': 'xlib'}],
@@ -1442,6 +1471,20 @@ check_types = [
   ['VdpPictureInfoAV1', ['vdpau/vdpau.h']],
 ]
 
+if nvenc_opt.allowed() and nvidia_codec_supported
+  check_types += [
+    ['NV_ENC_PIC_PARAMS_AV1', ['ffnvcodec/nvEncodeAPI.h'],
+      {'preconditions': ['ffnvcodec'], 'extra_deps': ['ffnvcodec_extra_deps']}],
+  ]
+endif
+
+if (nvdec_opt.allowed() or cuvid_opt.allowed()) and nvidia_codec_supported
+  check_types += [
+    ['CUVIDAV1PICPARAMS', ['ffnvcodec/dynlink_cuda.h', 'ffnvcodec/dynlink_cuviddec.h'],
+      {'preconditions': ['ffnvcodec'], 'extra_deps': ['ffnvcodec_extra_deps']}],
+  ]
+endif
+
 if get_option('vaapi').allowed()
   check_types += [
     ['VAPictureParameterBufferHEVC', ['va/va.h', 'va/va_dec_hevc.h']],
@@ -1763,6 +1806,7 @@ config_extra = [
   'cbs_vp8',
   'cbs_vp9',
   'celp_math',
+  'cuvidav1picparams',
   'd3d12_intra_refresh',
   'd3d12_motion_estimator',
   'd3d12_video_process_reference_info',
@@ -1827,6 +1871,7 @@ config_extra = [
   'msmpeg4dec',
   'msmpeg4enc',
   'mss34dsp',
+  'nv_enc_pic_params_av1',
   'pixblockdsp',
   'qpeldsp',
   'qsv',
@@ -1951,6 +1996,16 @@ check_compiles = [
   ['intrinsics_sse2', ['emmintrin.h'], '__m128i test = _mm_setzero_si128();'],
 ]
 
+if nvenc_opt.allowed() and nvidia_codec_supported
+  check_compiles += [
+    ['nvenc', ['ffnvcodec/nvEncodeAPI.h'], '''
+      NV_ENCODE_API_FUNCTION_LIST flist;
+      void f(void) { struct { const GUID guid; } codecs[] = { { NV_ENC_CODEC_H264_GUID } }; }
+      int main(void) { return 0; }
+    ''', {'preconditions': ['ffnvcodec'], 'extra_deps': ['ffnvcodec_extra_deps']}],
+  ]
+endif
+
 if get_option('schannel').allowed()
   check_compiles += [
     ['dtls_protocol', ['windows.h', 'security.h', 'schnlsp.h'], 'int i = SP_PROT_DTLS1_X_CLIENT;', {'c_args': ['-DSECURITY_WIN32'], 'link_with': ['secur32', 'ncrypt', 'crypt32']}],
@@ -2178,7 +2233,11 @@ if ['linux', 'openbsd', 'android', 'freebsd', 'netbsd'].contains(host_system)
   conf.set('section_data_rel_ro', 1)
 endif
 
-if get_option('ff_cuda_nvcc').allowed()
+use_cuda_nvcc = get_option('ff_cuda_nvcc').enable_auto_if(
+  get_option('ff_cuda_sdk').enabled() and not get_option('ff_cuda_llvm').enabled(),
+).enabled()
+
+if use_cuda_nvcc
   nvcc_default = find_program('nvcc')
   nvccflags_default = ['-gencode', 'arch=compute_30,code=sm_30', '-O2']
 
@@ -2199,7 +2258,7 @@ else
   nvccflags += ['-m32']
 endif
 
-if get_option('ff_cuda_nvcc').allowed()
+if use_cuda_nvcc
   nvccflags += ['-ptx']
 else
   nvccflags += ['-S', '-nocudalib', '-nocudainc', '--cuda-device-only', '-Wno-c++11-narrowing', '-include', meson.current_source_dir() / 'compat/cuda/cuda_runtime.h']
@@ -2216,7 +2275,7 @@ f = configure_file(
   capture: true,
 )
 
-if get_option('ff_cuda_nvcc').allowed()
+if use_cuda_nvcc
   cuda_llvm = run_command(nvcc_default, nvccflags, f, check: false)
   if cuda_llvm.returncode() != 0
     error('failed checking for nvcc.')
-- 
2.55.0

